### PR TITLE
[Parse] Avoid skipping bodies with `#sourceLocation`

### PR DESCRIPTION
**Pull request by hamishknight**
`#sourceLocation` requires setting and restoring state in the parser, so we cannot skip any bodies containing `#sourceLocation` tokens. Member skipping was already doing this, extend it to function body skipping too.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -7080,25 +7085,25 @@ bool Parser::canDelayMemberDeclParsing(bool &HasOperatorDeclarations,
...
   skipUntilMatchingRBrace(*this,
                           HasPoundDirective,
+                          HasPoundSourceLocation,
                           HasOperatorDeclarations,
                           HasNestedClassDeclarations,
                           HasNestedTypeDeclarations,
                           HasPotentialRegexLiteral);
-  if (!HasPoundDirective && !HasPotentialRegexLiteral)
+  if (!HasPoundDirective && !HasPotentialRegexLiteral) {
+    ASSERT(!HasPoundSourceLocation);
```
Comment (ahoppen) [inline lib/Parse/ParseDecl.cpp:7104]: Why can't `HasPoundSourceLocation` be true here?

Out of curiosity: Why are you using uppercase `ASSERT` here? Don’t we usually spell `assert` with lowercase?

Comment (ahoppen) [reply lib/Parse/ParseDecl.cpp:7104]: Ah, I see. Might be worth mentioning in a comment but not too important.

**Review by rintaro — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Pull request merged**